### PR TITLE
Change interface and add epsilon

### PR DESCRIPTION
**Pull request by martinhruska**
This PR:

- Changes default interface from result of function passed via parameter pointer to result returned by function directly by its return value. 
- Adds support for epsilon. Epsilon is the maximal value of symobl data type. It also add back() to OrdVector to enable accesing epsilon transitions faster

What is not included:
-  I did not move the functions returning results via parameter pointer to a special file. Actually, I didn't notice that this was approved on the last meeting and I don't think it is a clear solution. We would need to use friend function or create some kind of helping structure. I think neither one of the options is worth tradeoff for increasing entropy by creating new file. Moreover, any mediocre C++ programmer will understand why there are different signatures for the same function.

### CODE REVIEW

**Comment by martinhruska** [comment]
And I would need your help with update of Python interface. Thanks. @Adda0 @tfiedor

File: src/nfa/nfa-intersection.cc
Comment (Adda0) [inline src/nfa/nfa-intersection.cc:389]: You are now computing the same intersection twice. Return `intersection.get_product();` if you want to add product map construction like this, then.

File: include/mata/nfa.hh
Comment (Adda0) [inline include/mata/nfa.hh:238]: The `Limits` struct should be updated to use proper limits from `<limits>` header. We now use C limits and long max instead of unsigned long max at that. Furthermore, this constant should be then `constexpr` instead of `const`.

```cpp
static constexpr struct Limits {
    State maxState = std::numeric_limits<State>::max();
    State minState = std::numeric_limits<State>::min();
    Symbol maxSymbol = std::numeric_limits<Symbol>::max();
    Symbol minSymbol = std::numeric_limits<Symbol>::min()
} limits;
```

**Review by Adda0 — CHANGES_REQUESTED** [review]
LGTM in general. I agree with your decision to keep the functions in the same file.

We could also maybe use a separate `Mata::Nfa::utils` namespace (or something similar) for the pointer functions. What do you think about that?

File: src/nfa/nfa-intersection.cc
Comment (martinhruska) [reply src/nfa/nfa-intersection.cc:389]: My fault. I'm gonna fix it.

File: include/mata/nfa.hh
Comment (martinhruska) [reply include/mata/nfa.hh:236]: done

File: include/mata/nfa.hh
Comment (Adda0) [inline include/mata/nfa.hh:235]: Epsilon symbol can be made `constexpr` as well now.

File: include/mata/ord_vector.hh
Comment (Adda0) [inline include/mata/ord_vector.hh:479]: We need to get the element (with its transitions), not just `bool`.

Comment (Adda0) [reply include/mata/ord_vector.hh:479]: Furthermore, calling `back()` on an empty vector causes UB.

Either we have to always first check whether the given state has any transitions at all (our back would cause UB, too, following a behaviour of `std::vector<T>::back()`), or we must implement some logic handling this case.

We could return a pointer, which we would fill with `nullptr` in the case the vector is empty. We could also create a method `Mata::Nfa::Nfa::get_epsilon_transitions()`, which would return a `const_iterator` to `TransitionList` which would either be an iterator to the epsilon transitions, or `TransitionList::end()` in case there are no epsilon transitions for the given state.

I strongly recommend using the first option:
- follow the behaviour of `std::vector<T>::back()` with returning `T&` from `OrdVector::back()`,
- create method `get_epsilon_transitions()` which checks for emptiness of `TransitionList` and either returns `TransitionList::end()` if there are no epsilon transitions, or return a `const_iterator` to the epsilon transitions.

Comment (martinhruska) [reply include/mata/ord_vector.hh:479]: I agree. In OrdVector we should follow behaviour of underlying data structure. And add the `get_epsilon_transitions()` method to Nfa.

File: include/mata/nfa.hh
Comment (martinhruska) [reply include/mata/nfa.hh:846]: we can try.

File: include/mata/nfa.hh
Comment (Adda0) [inline include/mata/nfa.hh:828]: Error with automatic replacement of text. Revert this change.

File: include/mata/nfa.hh
Comment (Adda0) [reply include/mata/nfa.hh:846]: We can try to think of something better. However, an important problem here is the following: Error with automatic replacement of text. Revert this change.

File: include/mata/nfa.hh
Comment (Adda0) [inline include/mata/nfa.hh:855]: Error with automatic replacement of text. Revert this change.

File: include/mata/nfa.hh
Comment (Adda0) [inline include/mata/nfa.hh:865]: Error with automatic replacement of text. Revert this change.

File: include/mata/nfa.hh
Comment (Adda0) [inline include/mata/nfa.hh:867]: Error with automatic replacement of text. Revert this change.

File: include/mata/nfa.hh
Comment (Adda0) [inline include/mata/nfa.hh:915]: The default epsilon value should be specified here as well.

File: src/nfa/nfa-concatenation.cc
Comment (Adda0) [inline src/nfa/nfa-concatenation.cc:306]: Why do you remove the `const` qualifiers here? Once the epsilon is specified, it will remain `const` for the whole duration of the function execution.

File: src/nfa/nfa-intersection.cc
Comment (Adda0) [inline src/nfa/nfa-intersection.cc:51]: Error with automatic replacement of text. Revert this change.

File: src/nfa/nfa-intersection.cc
Comment (Adda0) [inline src/nfa/nfa-intersection.cc:371]: Why do you remove the const qualifiers here? Once the epsilon is specified, it will remain const for the whole duration of the function execution.

File: src/nfa/nfa.cc
Comment (Adda0) [inline src/nfa/nfa.cc:1261]: You can use `Mata::Nfa::Nfa::is_state(state)` for exactly this check.

File: src/nfa/nfa.cc
Comment (Adda0) [inline src/nfa/nfa.cc:1262]: To not making a copy of the transitions.
```suggestion
    const auto& trans_from = this->get_transitions_from(state);
```

**Review by Adda0 — CHANGES_REQUESTED** [review]
I am content with the current logic. However, I request to address the mentioned issues as well as the additional PR refactoring `get_epsilon_transitions()` method.

**Comment by Adda0** [comment]
First resolve #69 before merging this PR.

File: src/nfa/nfa-intersection.cc
Comment (martinhruska) [reply src/nfa/nfa-intersection.cc:371]: By mistake

File: src/nfa/nfa-concatenation.cc
Comment (martinhruska) [reply src/nfa/nfa-concatenation.cc:306]: mistake

File: include/mata/ord_vector.hh
Comment (Adda0) [reply include/mata/ord_vector.hh]: The logic seems OK. I have refactorized the function further in #69. Closing this conversation as resolved.

File: src/nfa/nfa-concatenation.cc
Comment (tfiedor) [reply src/nfa/nfa-concatenation.cc]: Your face is mistake.

**Review by Adda0 — APPROVED** [review]
LGTM now. Updates to the current algorithms using epsilon symbol will come with the next PR. C++ interface should be working as intended now. Another PR with python binding fixes will follow afterwards.

**Comment by tfiedor** [comment]
> LGTM now. Updates to the current algorithms using epsilon symbol will come with the next PR. C++ interface should be working as intended now. Another PR with python binding fixes will follow afterwards.

I think we should fix the binding (so it can be build, i.e., the action doesn't fail) first. I disaprove of merging with failed actions. The question is...who will fix the binding? :P Me or you?

**Comment by Adda0** [comment]
I would say that the `devel` branch is meant exactly for this kind of shenanigans. However, right now we are using `devel` branch as our master stable branch. Therefore, I completely agree that it would be better to fix the binding first. I can do that tomorrow in a separate PR merged into this branch. Does it work for you? Will my change to the few affected functions introduce any merge conflicts with your updates to Python binding?

**Comment by tfiedor** [comment]
> I would say that the `devel` branch is meant exactly for this kind of shenanigans. However, right now we are using `devel` branch as our master stable branch. Therefore, I completely agree that it would be better to fix the binding first. I can do that tomorrow morning in a separate PR to be merged into this branch. Does it work for you? Will my change to the few affected functions introduce any merge conflicts with your updates to Python binding?

I think, it is safe to push directly to this branch, without the need of separate PR. If Martin is finished with his work, there will not be any conflicts.

**Comment by Adda0** [comment]
I could do that, of course. (Even though I have pushes to the upstream repo turned off entirely for safety reasons.)

> there will not be any conflicts.

I was talking about yours (@tfiedor's) changes to the upcoming PR updating Python binding you have talked about in our meeting today. Whether I am likely to break something for you. Or, if that would not be a problem, you can then just rebase onto the updated `devel`.

**Comment by tfiedor** [comment]
> I could do that, of course. (Even though I have pushes to the upstream repo turned off entirely for safety reasons.)
> 
> > there will not be any conflicts.
> 
> I was talking about yours (@tfiedor's) changes to the upcoming PR updating Python binding you have talked about in our meeting today. Whether I am likely to break something for you. Or, if that would not be a problem, you can then just rebase onto the updated `devel`.

No worries, you will probably not have any conflicts with me. I will rebase my branch and solve conflicts (if any) myself.

**Comment by Adda0** [comment]
Do not merge for now, then. I will fix the Python binding first thing tomorrow morning (where morning means before noon :smile:).

File: include/mata/nfa.hh
Comment (Adda0) [reply include/mata/nfa.hh]: We can use the full name (as we already do in Python binding): `intersection_preserving_epsilon_transitions()`. Or, if you feel like this name is too long, its variations such as `intersection_pres_eps_trans()`. What do you think about these names? I would say they describe better what the operation does.

**Comment by Adda0** [comment]
I have pushed here the update to the Python binding. Feel free to take a look at the changes. As far as I am concerned, I consider this PR to be ready for merging.

**Pull request merged**
